pennylane/_grad.py

```diff
@@ -181,3 +182,206 @@ def _jacobian_function(*args, **kwargs):
+            )
+
+    i, j = idx
+
+    # diagonal
+    if i == j:
+        shift = _np.zeros_like(x)
+        shift[i] += delta
+        deriv2 = (
+            f(*args[:argnum], x + shift, *args[argnum + 1 :], **kwargs)
+            - 2 * f(*args[:argnum], x, *args[argnum + 1 :], **kwargs)
+            + f(*args[:argnum], x - shift, *args[argnum + 1 :], **kwargs)
+        ) * delta ** -2
+
+    # off-diagonal
+    if i != j:
+        shift_i = _np.zeros_like(x)
+        shift_i[i] += 0.5 * delta
+
+        shift_j = _np.zeros_like(x)
+        shift_j[j] += 0.5 * delta
+
+        deriv2 = (
+            f(*args[:argnum], x + shift_i + shift_j, *args[argnum + 1 :], **kwargs)
+            - f(*args[:argnum], x - shift_i + shift_j, *args[argnum + 1 :], **kwargs)
+            - f(*args[:argnum], x + shift_i - shift_j, *args[argnum + 1 :], **kwargs)
+            + f(*args[:argnum], x - shift_i - shift_j, *args[argnum + 1 :], **kwargs)
+        ) * delta ** -2
+
+    return deriv2
+
+
+def finite_diff(f, N=1, argnum=0, idx=None, delta=0.01):
+    r"""Returns a function that can be evaluated to compute the gradient or the
+    second-order derivative of the callable function ``f`` using a centered finite
+    difference approximation.
+
+    The first-order derivatives :math:`\frac{\partial f(x)}{\partial x_i}` entering
+    the gradient of the input function are given by,
+
+    .. math::
+
+        \frac{\partial f(x)}{\partial x_i} \approx \frac{f(x_i + \delta/2)
+        - f(x_i - \delta/2)}{\delta}
+
+    On the other hand, the second-order derivative
+    :math:`\frac{\partial^2 f(x)}{\partial x_i \partial x_j}` are evaluated using the
+    following expressions:
+
+    For :math:`i = j`
+
+    .. math::
+        \frac{\partial^2 f(x)}{\partial x_i^2} \approx
+        \frac{f(x_i + \delta) - 2 f(x) + f(x_i - \delta)}{\delta^2},
+
+    and for :math:`i \neq j`
+
+    .. math::
+        \frac{\partial^2 f(x)}{\partial x_i \partial x_j} \approx
+        \frac{f(x_i + \delta/2, x_j + \delta/2) - f(x_i - \delta/2, x_j + \delta/2)
+        - f(x_i + \delta/2, x_j - \delta/2) + f(x_i - \delta/2, x_j - \delta/2)}
+        {\delta^2}.
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        N (int): specifies the order of the finite difference approximation
+        argnum (int): the argument of function ``f`` to differentiate
+        idx (list[int]): If argument ``args[argnum]`` is an array, `idx`` can be used
+            to specify the indices of the argument ``argnum`` to differentiate.
+            For ``N=1`` it can be given to specify the gradient components to be computed.
+            For ``N=2``, it specifies the indices ``i, j`` of the variables involved in the
+            derivative :math:`\frac{\partial^2 f(x)}{\partial x_i \partial x_j}`.
+        delta (float): step size used to evaluate the finite differences
+
+    Returns:
```

Comment:
Yes, not finished yet.

Reply:
Agree, every user-facing function in PL should have an `**Example**` section after the arguments!

Reply:
An example of the full hessian might also be nice.

Reply:
Examples added.

@josh146, if it is OK with you, I would rather have an example for the Hessian in the docstring of the second-order derivative of the *energy* since this is something users would like to do in order to compute normal modes.